test/integration/test_celery_user_rate_limit.py

```diff
@@ -0,0 +1,169 @@
+    Test feature that limits the number of celery task executions
+    per user per second. This is implemented by the celery_user_rate_limit
+    config parameter whose default value is 0.0 meaning no rate limit.
+    For each of a set of users it submits runs a task num_calls times.
+    For the test to succeed the total duration of completion of all task
+    executions should be equal to the tasks_per_user_per_sec times
+    num_calls-1. This is because tasks for different users should be
+    executed in parallel. In addition, we verify that the total
+    duration for execution of tasks for each user equals the total
+    task duration, again because tasks for different users should
+    run in parallel.
+    """
+
+    def setUp(self):
+        super().setUp()
+
+    def _test_mock_pass_user_id_task(self, users: Iterable[int], num_calls: int, tasks_per_user_per_sec: float):
+        expected_duration: float
+        if tasks_per_user_per_sec == 0.0:
+            expected_duration = 0.0
+            expected_duration_lbound = 0.0
+        else:
+            secs_between_tasks_per_user = 1 / tasks_per_user_per_sec
+            expected_duration = secs_between_tasks_per_user * (num_calls - 1)
+            expected_duration_lbound = expected_duration - 4
+        expected_duration_hbound = expected_duration + 4
+        start_time = datetime.datetime.utcnow()
+        timer = ExecutionTimer()
+        results: Dict[int, List[AsyncResult]] = {}
+        for user in users:
+            user_results: List[AsyncResult] = []
+            for _i in range(num_calls):  # type: ignore
+                user_results.append(mock_user_id_task.delay(task_user_id=user))
+            results[user] = user_results
+        for user, user_results in results.items():
+            for result in user_results:
+                val = result.get(timeout=1000)
+                assert val == user
+        elapsed = timer.elapsed
+        assert elapsed >= expected_duration_lbound and elapsed <= expected_duration_hbound
+        for user_results in results.values():
+            last_task_end_time = start_time
+            for result in user_results:
+                if result.date_done > last_task_end_time:
+                    last_task_end_time = result.date_done
+            user_elapsed = (last_task_end_time - start_time).total_seconds()
+            assert user_elapsed >= expected_duration_lbound and user_elapsed <= expected_duration_hbound
+
+
+class TestCeleryUserRateLimitIntegrationPostgres(TestCeleryUserRateLimitIntegration):
+    @classmethod
+    def handle_galaxy_config_kwds(cls, config):
+        dburl = config["database_connection"]
+        setup_users(dburl)
+
+
+class TestCeleryUserRateLimitIntegrationPostgres1(TestCeleryUserRateLimitIntegrationPostgres):
+    @classmethod
+    def handle_galaxy_config_kwds(cls, config):
+        TestCeleryUserRateLimitIntegrationPostgres.handle_galaxy_config_kwds(config)
+        config["celery_user_rate_limit"] = 0.1
```

Comment:
Since each test class sets the rate to a value and then passes that value to the test function, could you use a class variable instead? Easier to read + less error-prone + immediately clear what's different about that test.

Reply:
ok